File: Code/1991.cpp
#include <stdio.h>
#include <iostream>

using namespace std;

int arr[26][2];

//root-left-right
void preorder(int node){

        if (node == -1) return;

        else {
                cout << char(node + 'A');
                preorder(arr[node][0]);
                preorder(arr[node][1]);

        }

}

//left-root-right
void inorder(int node){

        if (node == -1) return;

        else {
                inorder(arr[node][0]);
                cout << char(node + 'A');
                inorder(arr[node][1]);

        }
}

//left-right-root
void postorder(int node){

        if (node == -1) return;

        else {
                postorder(arr[node][0]);
                postorder(arr[node][1]);
                cout << char(node + 'A');

        }
}

int main(void){

        int N;

        scanf("%d", &N);

        for( int i = 0 ; i < N ; i++ ){
                char root, L, R ; 

                cin >> root >> L >> R;
                
                if (L != '.') arr[root-'A'][0] = L-'A';
                else arr[root-'A'][0] = -1;

                if (R != '.') arr[root-'A'][1] = R-'A';
                else arr[root-'A'][1] = -1;


        }
        preorder(0);
        printf("\n");
        inorder(0);
        printf("\n");
        postorder(0);
        printf("\n");

        return 0;
}

File: Code/python/algorithm/tree/1991.py

# root-left-right
def preorder(node, tree):
    if node == '.' :
        return
    print(node, end = '')
    preorder(tree[node][0], tree)
    preorder(tree[node][1], tree)

# left-root-right
def inorder(node, tree):
    if node == '.' :
        return
    inorder(tree[node][0], tree)
    print(node, end = '')
    inorder(tree[node][1], tree)
    
# left-right-root
def postorder(node, tree):
    if node == '.' : 
        return
    postorder(tree[node][0], tree)
    postorder(tree[node][1], tree)
    print(node, end = '')

if __name__ == '__main__' : 

    n = int(input())
    tree = {}

    for _ in range(n):
        root, left, right = input().split()
        tree[root] = (left, right)

    preorder('A', tree)
    print()
    inorder('A', tree)
    print()
    postorder('A', tree)
    print()